Issue #2591177: Rules definition objects are almost aways returning da…

torenware opened the pull request:
…ta type 'any'.

This is pretty preliminary, but I wanted this to be where folks can find it.

Drupal Issue https://www.drupal.org/node/2591177

torenware commented:
This one is no-go; looks like I broke the deriver.  When I figure this out, I'll push an update.

torenware commented:
It turns out my fix was fine.  The reasons the tests broke is that until now, our mocks have only handled ContextDefinitions with a DataType of 'any', so our entity type mocks needed to be expanded to cover the types of entity:user, entity:user_role, and entity:node. This is in addition to the entity:test type that we were already mocking.

torenware commented:
Per @klausi, I've added an integration test to test whether Rules plugins correctly report their data type.

klausi commented on tests/src/Integration/Engine/AnnotationProcessingTest.php:
```diff
@@ -0,0 +1,136 @@
+    $this->assertNotNull($plugin, "{$params['type']} plugin {$params['plugin']} loads");
+    if (empty($plugin)) {
+      return;
+    }
```
why is that if() necessary? I think phpunit already stops when an assertion fails.

klausi commented on tests/src/Integration/Engine/AnnotationProcessingTest.php:
```diff
@@ -0,0 +1,136 @@
+   *
+   * @group rules
+   */
+  public function testCheckConfiguration($params) {
```
instead of $params can we separate out all 4 parameters to make the test easier to read?

klausi commented on tests/src/Integration/Engine/AnnotationProcessingTest.php:
```diff
@@ -0,0 +1,136 @@
+    $this->assertNotNull($context, "Plugin {$params['plugin']} has context {$params['context']}");
+    if (empty($context)) {
+      return;
+    }
```
another if() that can be removed.

klausi commented on tests/src/Integration/Engine/AnnotationProcessingTest.php:
```diff
@@ -0,0 +1,136 @@
+use Drupal\Core\Session\SessionManagerInterface;
+
+/**
+ * Test auto saving of variables after Rules execution.
```
Comment is wrong

torenware replied:
Fixed.

klausi replied:
forgot to push?

torenware replied:
Nope, I pushed it.  If you view the file with the View button, you'll see I made the fix.  For some reason, though, the diff shows the old state of the file.  Github bug I'm guessing.

klausi commented on tests/src/Integration/RulesEntityIntegrationTestBase.php:
```diff
@@ -53,15 +53,47 @@ public function setUp() {
+    $type_array = [];
+
+    foreach ($type_info as $type => $info) {
+      $entity_type = new ContentEntityType($info);
```
a user role is not a content entity type, so this does not really fit.

torenware replied:
We do need a mock for it, so I'll make sure that it's passed as a ConfigEntityType.

klausi commented on src/Context/ContextDefinition.php:
```diff
@@ -81,7 +81,12 @@ public static function createFromArray($values) {
+      }
+      else {
+        $values['type'] = $values['value'];
+      }
```
why do we have to do this? Can't we just change $nameMap and point 'value' => 'dataType'? Or are we using 'type' as key anywhere else?

torenware replied:
@klausi -- yeah, we are using values['type'] down stream. I'm can find you the references if you want.  Changing them from 'type' to 'value' should probably be a separate issue, however.

torenware commented:
OK, looks like I fixed the wrong comment.  Fixed both now.

torenware commented on tests/src/Integration/RulesEntityIntegrationTestBase.php:
```diff
@@ -53,15 +54,48 @@ public function setUp() {
+      'label' => 'Test Role',
+    ];
+    $role_type = new ConfigEntityType($role_entity_info);
+    $type_array['user_role'] = $role_type;
```
Note that the role is now handled as a ConfigEntityType.

klausi commented on tests/src/Integration/Engine/AnnotationProcessingTest.php:
```diff
@@ -0,0 +1,126 @@
+  public function setUp() {
+    parent::setUp();
+    $this->enableModule('user');
+    $this->sessionManager = $this->prophesize(SessionManagerInterface::class);
```
all class variables should be declared and documented on the class. And why would an annotation test need a session manager? Please add a comment.

torenware replied:
The issue here is that I'm loading a set of plugins and testing the annotations as used in those plugins.  If a plugin has an entity:user context, the plugin will invoke APIs that will require a Session Manager mock.  

When constructing the mocks, I was letting Prophecy tell me what methods of what interfaces it required.  It apparently needed SessionMangerInterface available in services container.

I'll check why $this->sessionManager is not declared here, since I'd agree that it needs to be there if I refer to it.  It may not be necessary, however, since I suspect that adding the service to the container will suffice.

torenware committed:
Issue #2591177: Rules definition objects are almost ways returning data type 'any'.

 - Fixed handling of type in our modified ContextDefinition annotation handler.
 - Added mocks for multiple entity types, since until now, all our ContextDefinitions
   have been of type 'any'.
 - Added a test that demonstrates the problem.

klausi commented:
merged, thanks!
